plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/TableStatisticsWriter.java

```diff
@@ -149,6 +178,73 @@ public StatisticsFile writeStatisticsFile(
         }
     }
 
+    private CollectedStatistics mergeStatisticsIfNecessary(
+            Table table,
+            long snapshotId,
+            FileIO fileIO,
+            StatsUpdateMode updateMode,
+            CollectedStatistics collectedStatistics)
+    {
+        if (updateMode == INCREMENTAL_UPDATE) {
+            Snapshot snapshot = table.snapshot(snapshotId);
+            checkState(snapshot != null, "No snapshot information for snapshotId %s in table %s", snapshotId, table);
+            if (snapshot.parentId() == null || !maySnapshotHaveData(table, snapshot.parentId(), fileIO)) {
+                // No previous snapshot, or previous snapshot empty
+                updateMode = REPLACE;
+            }
+        }
+
+        switch (updateMode) {
+            case REPLACE -> {
+                return collectedStatistics;
+            }
+
+            case INCREMENTAL_UPDATE -> {
+                Map<Integer, CompactSketch> collectedNdvSketches = collectedStatistics.ndvSketches();
+                ImmutableMap.Builder<Integer, CompactSketch> ndvSketches = ImmutableMap.builder();
+
+                Set<Integer> pendingPreviousNdvSketches = new HashSet<>(collectedNdvSketches.keySet());
+                Iterator<StatisticsFile> statisticsFiles = walkStatisticsFiles(table, snapshotId);
+                while (!pendingPreviousNdvSketches.isEmpty() && statisticsFiles.hasNext()) {
+                    StatisticsFile statisticsFile = statisticsFiles.next();
+
+                    boolean hasUsefulData = statisticsFile.blobMetadata().stream()
+                            .filter(blobMetadata -> blobMetadata.type().equals(StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1))
+                            .filter(blobMetadata -> blobMetadata.fields().size() == 1)
+                            .anyMatch(blobMetadata -> pendingPreviousNdvSketches.contains(getOnlyElement(blobMetadata.fields())));
+
+                    if (hasUsefulData) {
```

Comment:
nit: should you assert that what you actually read from puffin file matches what was in the metadata?

Reply:
I don't think we need to cross-check puffin vs metadata. I assume the table isn't corrupted.

BTW i was thinking that maybe we shouldn't traverse all the stat files.
Instead, we could require that the most recent stat file has all the useful statistics.
(this would be a change on read & write path, ie also in `io.trino.plugin.iceberg.TableStatisticsReader#readNdvs`)

@alexjo2144 @losipiuk @electrum thoughts?
cc @rdblue

Reply:
In a situation like:
- user analyzes columns c1, c2, c3
- user inserts some data
- user realizes s3 was not helpful, reanalyzes c1 and c2
- user inserts some data

As is, would we keep the drifting stats for c3, or would they be dropped?

Reply:
as a followup i logged this

> BTW i was thinking that maybe we shouldn't traverse all the stat files.